### PR TITLE
Build from source

### LINKED ISSUE #16: Build from source

### PR DESCRIPTION
**Pull request by proletarius101**
Fixes #16

### CODE REVIEW

**Comment by proletarius101** [comment]
bot, build com.bitwarden.desktop

**Comment by proletarius101** [comment]
Any idea why this would happen? @ghisvail 
````
[3 preload-host-spawn-strategy] Running: /app/bin/zypak-helper child - /app/bitwarden/bitwarden --type=zygote
A JavaScript error occurred in the main process
Uncaught Exception:
Error: The module '/tmp/.org.chromium.Chromium.HRJcMo'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 83. This version of Node.js requires
NODE_MODULE_VERSION 85. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
    at process.func [as dlopen] (electron/js2c/asar_bundle.js:5:1812)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1203:18)
    at Object.func [as .node] (electron/js2c/asar_bundle.js:5:2039)
    at Module.load (internal/modules/cjs/loader.js:992:32)
    at Module._load (internal/modules/cjs/loader.js:885:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12738)
    at Module.require (internal/modules/cjs/loader.js:1032:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/app/bitwarden/resources/app.asar/node_modules/desktop-idle/index.js:1:21)
    at Module._compile (internal/modules/cjs/loader.js:1152:30)
````

**Comment by ghisvail** [comment]
Maybe try with node12 instead of node14. I believe this is still the version used upstream for packaging.

**Comment by proletarius101** [comment]
> node12

Tried with node 10, 12, and 14 locally with no luck, but it could also be a problem of my cache. Let's see whether the CI will build without the issue.

**Comment by proletarius101** [comment]
Unfortunately, no luck :(

**Comment by proletarius101** [comment]
Wait, probably because I didn't change the append-path... Let me fix it

**Comment by proletarius101** [comment]
Maybe better, but
````
A JavaScript error occurred in the main process
Uncaught Exception:
Error: /tmp/.org.chromium.Chromium.tTdVkp: undefined symbol: _ZN2v816FunctionTemplate3NewEPNS_7IsolateEPFvRKNS_20FunctionCallbackInfoINS_5ValueEEEENS_5LocalIS4_EENSA_INS_9SignatureEEEiNS_19ConstructorBehaviorENS_14SideEffectTypeE
    at process.func [as dlopen] (electron/js2c/asar_bundle.js:5:1812)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1203:18)
    at Object.func [as .node] (electron/js2c/asar_bundle.js:5:2039)
    at Module.load (internal/modules/cjs/loader.js:992:32)
    at Module._load (internal/modules/cjs/loader.js:885:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12738)
    at Module.require (internal/modules/cjs/loader.js:1032:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/app/bitwarden/resources/app.asar/node_modules/desktop-idle/index.js:1:21)
    at Module._compile (internal/modules/cjs/loader.js:1152:30)
````

**Comment by proletarius101** [comment]
https://github.com/nodejs/node-v0.x-archive/issues/25483 suggests that it's because of v8's signature change.
````
make: Entering directory '/run/build/bitwarden/build/node_modules/desktop-idle/build'
  ccache g++ '-DNODE_GYP_MODULE_NAME=desktopIdle' '-DUSING_UV_SHARED=1' '-DUSING_V8_SHARED=1' '-DV8_DEPRECATION_WARNINGS=1' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/usr/lib/sdk/node10/include/node -I/usr/lib/sdk/node10/src -I/usr/lib/sdk/node10/deps/openssl/config -I/usr/lib/sdk/node10/deps/openssl/openssl/include -I/usr/lib/sdk/node10/deps/uv/include -I/usr/lib/sdk/node10/deps/zlib -I/usr/lib/sdk/node10/deps/v8/include  -fPIC -pthread -Wall -Wextra -Wno-unused-parameter -m64 -O3 -fno-omit-frame-pointer -fno-rtti -fno-exceptions -std=gnu++1y -MMD -MF ./Release/.deps/Release/obj.target/desktopIdle/src/desktop_idle.o.d.raw  -O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -c -o Release/obj.target/desktopIdle/src/desktop_idle.o ../src/desktop_idle.cc
In file included from ../src/desktop_idle.cc:1:
/usr/lib/sdk/node10/include/node/node.h:573:7: warning: cast between incompatible function types from ‘void (*)(v8::Local<v8::Object>)’ to ‘node::addon_register_func’ {aka ‘void (*)(v8::Local<v8::Object>, v8::Local<v8::Value>, void*)’} [-Wcast-function-type]
  573 |       (node::addon_register_func) (regfunc),                          \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
````

**Comment by proletarius101** [comment]
So everything is good now

**Comment by proletarius101** [comment]
@ghisvail So any updates?

**Comment by TheEvilSkeleton** [comment]
@proletarius101 I've opened https://github.com/flathub/com.bitwarden.desktop/pull/54, since it's the new version now. You can bump to the latest version with this PR and I won't mind closing my PR.

**Comment by TheEvilSkeleton** [comment]
If @ghisvail isn't here then maybe we might need to ping the Flathub maintainers to push.

**Comment by ghisvail** [comment]
> If @ghisvail isn't here then maybe we might need to ping the Flathub maintainers to push.

Or be patient, and wait.

**Comment by ghisvail** [comment]
I'd be keen to see the changes required to update this PR to the latest version.

Maintaining build from source for Electron apps is notoriously painful (having done it in the past) and I still need to convince myself it's worth coming back to it despite the significant overhead of maintenance.

**Comment by proletarius101** [comment]
> Maintaining build from source for Electron apps is notoriously painful (having done it in the past) and I still need to convince myself it's worth coming back to it despite the significant overhead of maintenance.

Maybe you will like to use github actions to automatically update the sources like this
https://github.com/flathub/org.standardnotes.standardnotes/actions/runs/648285763. And that's what I'm trying to add to this PR. With a separate commit of course.

**Comment by TheEvilSkeleton** [comment]
> If @ghisvail isn't here then maybe we might need to ping the >> Flathub maintainers to push.
> 
> 
> Or be patient, and wait.

Sorry about that. I just think that the maintainer should be informing the submitter about their reasoning rather than not commenting anything.

**Comment by proletarius101** [comment]
Things left:
- https://github.com/bitwarden/desktop/issues/773
- metainfo should be automatically updated

**Comment by ghisvail** [comment]
> I'd be keen to see the changes required to update this PR to the latest version.

Sadly, I can't evaluate the changes because of the force-push.

**Comment by proletarius101** [comment]
> Sadly, I can't evaluate the changes because of the force-push.

It's still WIP. But sounds like you prefer not to force-push (squash commits). Then I'll do so accordingly

**Comment by ghisvail** [comment]
Well, not immediately. I would like to measure the necessary changes to
update the manifest from one software release to the next.

Le lun. 15 mars 2021 à 07:47, proletarius101 ***@***.***> a
écrit :

**Comment by proletarius101** [comment]
So looks like the metainfi IS updated by the flathubbot already: https://github.com/flathub/org.standardnotes.standardnotes/pull/6/files

File: com.bitwarden.desktop.desktop [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[Desktop Entry]
+Name=Bitwarden
+Exec=start-standardnotes %u
```
[the remark this answers is not in the document]
Comment (proletarius101) [reply com.bitwarden.desktop.desktop]: Hmmm... Did I even leave it here? Looks like you reviewed a pretty old commit?

File: com.bitwarden.desktop.yaml [hunk 2]
```diff
@@ -19,29 +21,68 @@ rename-desktop-file: bitwarden.desktop
 rename-icon: bitwarden
 modules:
   - shared-modules/libsecret/libsecret.json
+  - shared-modules/python2.7/python-2.7.json
```
[the remark this answers is not in the document]
Comment (proletarius101) [reply com.bitwarden.desktop.yaml]: No, and this change is outdated

File: flatpak-node-generator.py [hunk 3]
```diff
@@ -0,0 +1,1675 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (proletarius101) [reply flatpak-node-generator.py]: Outdated commit

**Comment by proletarius101** [comment]
Due to some unfortunate situations, I have to force push again. Good news is the previous changes are not reviewed

**Comment by ghisvail** [comment]
Go ahead :-)

Le mer. 17 mars 2021 à 11:40, proletarius101 ***@***.***> a
écrit :

**Comment by proletarius101** [comment]
Close in favor of https://github.com/flathub/com.bitwarden.desktop/pull/55

**Pull request closed without merging**